w3lib/url.py

```diff
@@ -204,8 +204,15 @@ def _add_or_replace_parameters(url, params):
     parsed = urlsplit(url)
     args = parse_qsl(parsed.query, keep_blank_values=True)
 
-    new_args = OrderedDict(args)
-    new_args.update(params)
+    new_args = []
+    changing_params = list(params.keys())
+    for name, value in args:
+        if name in params.keys():
+            new_args.append((name, params[name]))
+            del params[name]
```

Comment:
I don't think mutating `params` is a good idea, I'd prefer to flag seen params in another set and ignore them on line 215.

Reply:
Something in the line of 

```python
  def _add_or_replace_parameters(url, params):
      parsed = urlsplit(url)
      args = parse_qsl(parsed.query, keep_blank_values=True)
      acc = []
      seen = set()
      for k, v in args:
          if k in params:
              if k not in seen:
                  acc.append((k, params[k]))
              seen.add(k)
          else:
              acc.append((k, v))

      acc += ((k, v) for k, v in params.items() if k not in seen)
      query = urlencode(acc)
      return urlunsplit(parsed._replace(query=query))
```

Reply:
> I don't think mutating `params` is a good idea, I'd prefer to flag seen params in another set and ignore them on line 215.

As the input argument is a dict, it makes sense that it shouldn't be modified by the function. I included a test case to avoid this scenario as well.